include/fiction/algorithms/simulation/sidb/critical_temperature.hpp

```diff
@@ -0,0 +1,495 @@
+     * @return State type (i.e. transparent, erroneous) of the ground state is returned.
+     */
+    bool energy_between_ground_state_and_first_erroneous(const sidb_energy_and_state_type& energy_and_state_type,
+                                                         const double                      min_energy)
+    {
+        bool ground_state_is_transparent = false;
+        for (const auto& [energy, state_type] : energy_and_state_type)
+        {
+            // Check if there is at least one ground state that satisfies the logic (transparent). Round the energy
+            // value of the given valid_layout to six decimal places to overcome possible rounding errors and for
+            // comparability with the min_energy.
+            if ((round_to_n_decimal_places(energy, 6) == round_to_n_decimal_places(min_energy, 6)) && state_type)
+            {
+                ground_state_is_transparent = true;
+            }
+
+            if (!state_type && (energy > min_energy) && ground_state_is_transparent)
+            {
+                // The energy difference is stored in meV.
+                temperature_stats.energy_between_ground_state_and_first_erroneous = (energy - min_energy) * 1000;
+                break;
+            }
+        }
+        return ground_state_is_transparent;
+    };
+    /**
+     * The Critical Temperature is determined.
+     *
+     * @param energy_state_type All energies of all physically valid charge distributions with the corresponding state
+     * type (i.e. transparent, erroneous).
+     */
+    void determine_critical_temperature(const sidb_energy_and_state_type& energy_state_type)
+    {
+        // Vector with temperature values from 0.01 to max_temperature * 100 K in 0.01 K steps is generated.
+        std::vector<double> temp_values{};
+        temp_values.reserve(parameter.max_temperature * 100);
+
+        for (uint64_t i = 1; i <= parameter.max_temperature * 100; i++)
+        {
+            temp_values.push_back(static_cast<double>(i) / 100.0);
+        }
+        // This function determines the Critical Temperature for a given confidence level.
+        for (const auto& temp : temp_values)
+        {
+            // If the occupation probability of erroneous states exceeds the given threshold...
+            if (occupation_probability_gate_based(energy_state_type, temp) > (1 - parameter.confidence_level))
+            {
+                // The current temperature is stored as Critical Temperature.
+                temperature_stats.critical_temperature = temp;
+
+                break;
+            }
+
+            if (std::abs(temp - static_cast<double>(parameter.max_temperature)) < 0.001)
+            {
+                // Maximal temperature is stored as Critical Temperature.
+                temperature_stats.critical_temperature = static_cast<double>(parameter.max_temperature);
+            }
+        }
+    }
+
+    /**
+     * SiDB cell-level layout.
+     */
+    Lyt& layout{};
```

Comment:
I think for now it can be ``const``.  However, when using ``QuickExact`` it should be without ``const``. But let's go with ``const`` for now. Thanks for pointing this out!